include/deal.II/lac/petsc_matrix_base.h

```diff
@@ -1476,28 +1431,49 @@ namespace PETScWrappers
                   const bool         elide_zero_values,
                   const bool /*col_indices_are_sorted*/)
   {
-    (void)elide_zero_values;
+    add_or_set(VectorOperation::add,
+               row,
+               n_cols,
+               col_indices,
+               values,
+               elide_zero_values);
+  }
+
 
-    prepare_action(VectorOperation::add);
+
+  inline void
+  MatrixBase::add_or_set(const VectorOperation::values &operation,
+                         const size_type                row,
+                         const size_type                n_cols,
+                         const size_type               *col_indices,
+                         const PetscScalar             *values,
+                         const bool                     elide_zero_values)
+  {
+    prepare_action(operation);
 
     const auto petsc_row = static_cast<PetscInt>(row);
     AssertIntegerConversion(petsc_row, row);
 
-    PetscInt n_columns = 0;
-    column_indices.resize(n_cols);
-    column_values.resize(n_cols);
+    std::vector<PetscInt> &column_indices = scratch_column_indices.get();
+    column_indices.resize(0);
+    column_indices.reserve(n_cols);
```

Comment:
I don't quite see why you insist on re-allocating memory. Why not just do `resize()` in line 1467 (new) and clear()/reserve() in the `else` block after line 1478?

Reply:
That's probably more reasonable. My original thought was that `resize()` won't reallocate and we should set the correct size once we know it.

Based on the discussion I had with Matthias, though - TLS is not the right solution here so this will be replaced with `boost::small_vector` in about 10 minutes so it will be fixed in a different way.